File: src/pieces.rs
//! Piece types, asset names, and move generation.

use crate::board::{self, Board, in_bounds};
use crate::constants::{CastlingRights, SQUARES};
use std::collections::HashSet;

/// Asset filename for a piece. char: R,N,B,Q,K,P (black) or r,n,b,q,k,p (white).
pub fn asset_name(piece: char) -> Option<&'static str> {
    use crate::constants::*;
    let s = match piece {
        'r' => ASSET_ROOK_WHITE,
        'n' => ASSET_KNIGHT_WHITE,
        'b' => ASSET_BISHOP_WHITE,
        'q' => ASSET_QUEEN_WHITE,
        'k' => ASSET_KING_WHITE,
        'p' => ASSET_PAWN_WHITE,
        'R' => ASSET_ROOK_BLACK,
        'N' => ASSET_KNIGHT_BLACK,
        'B' => ASSET_BISHOP_BLACK,
        'Q' => ASSET_QUEEN_BLACK,
        'K' => ASSET_KING_BLACK,
        'P' => ASSET_PAWN_BLACK,
        _ => return None,
    };
    Some(s)
}

/// Generate all pseudo-legal target squares for the piece at (col, row).
/// Does not filter for check. Uses castling_rights for king castling; en_passant_target for en passant.
pub fn pseudo_legal_moves(
    board: &Board,
    col: i32,
    row: i32,
    castling_rights: &CastlingRights,
    en_passant_target: Option<(i32, i32)>,
) -> HashSet<(i32, i32)> {
    let piece = board::get_piece(board, col, row);
    if piece == ' ' {
        return HashSet::new();
    }
    let white = board::is_white_piece(piece);
    let mut targets = HashSet::new();

    match piece.to_ascii_lowercase() {
        'p' => pawn_moves(board, col, row, white, en_passant_target, &mut targets),
        'r' => sliding_moves(board, col, row, white, &[(0, 1), (0, -1), (1, 0), (-1, 0)], &mut targets),
        'b' => sliding_moves(board, col, row, white, &[(1, 1), (1, -1), (-1, 1), (-1, -1)], &mut targets),
        'q' => {
            sliding_moves(board, col, row, white, &[(0, 1), (0, -1), (1, 0), (-1, 0)], &mut targets);
            sliding_moves(board, col, row, white, &[(1, 1), (1, -1), (-1, 1), (-1, -1)], &mut targets);
        }
        'k' => {
            king_moves(board, col, row, white, &mut targets);
            king_castling_moves(board, col, row, white, castling_rights, &mut targets);
        }
        'n' => knight_moves(board, col, row, white, &mut targets),
        _ => {}
    }
    targets
}

fn pawn_moves(
    board: &Board,
    col: i32,
    row: i32,
    white: bool,
    en_passant_target: Option<(i32, i32)>,
    out: &mut HashSet<(i32, i32)>,
) {
    let forward = if white { -1 } else { 1 };
    let start_row = if white { 6 } else { 1 };

    // One square forward
    let r1 = row + forward;
    if in_bounds(col, r1) && board::is_empty(board, col, r1) {
        out.insert((col, r1));
        // Two squares from start
        if row == start_row {
            let r2 = row + 2 * forward;
            if in_bounds(col, r2) && board::is_empty(board, col, r2) {
                out.insert((col, r2));
            }
        }
    }
    // Captures (normal and en passant)
    for dc in [-1, 1] {
        let c = col + dc;
        let r = row + forward;
        if !in_bounds(c, r) {
            continue;
        }
        let p = board::get_piece(board, c, r);
        if p != ' ' && board::is_black_piece(p) == white {
            out.insert((c, r));
        } else if en_passant_target == Some((c, r)) {
            out.insert((c, r));
        }
    }
}

fn sliding_moves(
    board: &Board,
    col: i32,
    row: i32,
    white: bool,
    directions: &[(i32, i32)],
    out: &mut HashSet<(i32, i32)>,
) {
    for &(dc, dr) in directions {
        let mut c = col + dc;
        let mut r = row + dr;
        while in_bounds(c, r) {
            let p = board::get_piece(board, c, r);
            if p == ' ' {
                out.insert((c, r));
            } else {
                if board::is_black_piece(p) == white {
                    out.insert((c, r));
                }
                break;
            }
            c += dc;
            r += dr;
        }
    }
}

fn king_moves(board: &Board, col: i32, row: i32, white: bool, out: &mut HashSet<(i32, i32)>) {
    for dc in -1..=1 {
        for dr in -1..=1 {
            if dc == 0 && dr == 0 {
                continue;
            }
            let c = col + dc;
            let r = row + dr;
            if !in_bounds(c, r) {
                continue;
            }
            let p = board::get_piece(board, c, r);
            if p == ' ' || (board::is_black_piece(p) == white) {
                out.insert((c, r));
            }
        }
    }
}

/// True if (col, row) is attacked by any piece of the given side.
fn is_square_attacked(board: &Board, col: i32, row: i32, attacker_white: bool) -> bool {
    for r in 0..SQUARES {
        for c in 0..SQUARES {
            if !board::piece_is_side(board, c, r, attacker_white) {
                continue;
            }
            let piece = board::get_piece(board, c, r);
            let lower = piece.to_lowercase().next().unwrap_or(piece);
            let moves = match lower {
                'p' => {
                    let mut s = HashSet::new();
                    pawn_moves(board, c, r, attacker_white, None, &mut s);
                    s
                }
                'r' => {
                    let mut s = HashSet::new();
                    sliding_moves(board, c, r, attacker_white, &[(0, 1), (0, -1), (1, 0), (-1, 0)], &mut s);
                    s
                }
                'b' => {
                    let mut s = HashSet::new();
                    sliding_moves(board, c, r, attacker_white, &[(1, 1), (1, -1), (-1, 1), (-1, -1)], &mut s);
                    s
                }
                'q' => {
                    let mut s = HashSet::new();
                    sliding_moves(board, c, r, attacker_white, &[(0, 1), (0, -1), (1, 0), (-1, 0)], &mut s);
                    sliding_moves(board, c, r, attacker_white, &[(1, 1), (1, -1), (-1, 1), (-1, -1)], &mut s);
                    s
                }
                'k' => {
                    let mut s = HashSet::new();
                    king_moves(board, c, r, attacker_white, &mut s);
                    s
                }
                'n' => {
                    let mut s = HashSet::new();
                    knight_moves(board, c, r, attacker_white, &mut s);
                    s
                }
                _ => HashSet::new(),
            };
            if moves.contains(&(col, row)) {
                return true;
            }
        }
    }
    false
}

/// Add castling moves for the king at (col, row). King must be on e-file (col 4), white on row 7, black on row 0.
/// Rules: (1) King and the rook must not have moved (enforced via castling_rights).
/// (2) No pieces between king and rook. (3) King not in check, and does not pass through or land in check.
fn king_castling_moves(
    board: &Board,
    col: i32,
    row: i32,
    white: bool,
    rights: &CastlingRights,
    out: &mut HashSet<(i32, i32)>,
) {
    if col != 4 {
        return;
    }
    let opponent_white = !white;
    let (kingside_ok, queenside_ok) = if white {
        if row != 7 {
            return;
        }
        (
            rights.white_kingside,
            rights.white_queenside,
        )
    } else {
        if row != 0 {
            return;
        }
        (
            rights.black_kingside,
            rights.black_queenside,
        )
    };

    if kingside_ok {
        if board::is_empty(board, 5, row) && board::is_empty(board, 6, row)
            && board::get_piece(board, 7, row) == if white { 'r' } else { 'R' }
            && !is_square_attacked(board, 4, row, opponent_white)
            && !is_square_attacked(board, 5, row, opponent_white)
            && !is_square_attacked(board, 6, row, opponent_white)
        {
            out.insert((6, row));
        }
    }
    if queenside_ok {
        if board::is_empty(board, 1, row) && board::is_empty(board, 2, row) && board::is_empty(board, 3, row)
            && board::get_piece(board, 0, row) == if white { 'r' } else { 'R' }
            && !is_square_attacked(board, 4, row, opponent_white)
            && !is_square_attacked(board, 3, row, opponent_white)
            && !is_square_attacked(board, 2, row, opponent_white)
        {
            out.insert((2, row));
        }
    }
}

fn knight_moves(board: &Board, col: i32, row: i32, white: bool, out: &mut HashSet<(i32, i32)>) {
    let deltas: [(i32, i32); 8] = [
        (2, 1), (2, -1), (-2, 1), (-2, -1),
        (1, 2), (1, -2), (-1, 2), (-1, -2),
    ];
    for (dc, dr) in deltas {
        let c = col + dc;
        let r = row + dr;
        if !in_bounds(c, r) {
            continue;
        }
        let p = board::get_piece(board, c, r);
        if p == ' ' || (board::is_black_piece(p) == white) {
            out.insert((c, r));
        }
    }
}

/// Filter pseudo-legal moves to only those that don't leave the current side's king in check.
pub fn legal_moves(
    board: &Board,
    col: i32,
    row: i32,
    castling_rights: &CastlingRights,
    en_passant_target: Option<(i32, i32)>,
) -> HashSet<(i32, i32)> {
    let piece = board::get_piece(board, col, row);
    if piece == ' ' {
        return HashSet::new();
    }
    let white = board::is_white_piece(piece);
    let pseudo = pseudo_legal_moves(board, col, row, castling_rights, en_passant_target);
    let mut legal = HashSet::new();
    for (tc, tr) in pseudo {
        let mut b = *board;
        b[tr as usize][tc as usize] = piece;
        b[row as usize][col as usize] = ' ';
        if en_passant_target == Some((tc, tr)) {
            let capture_row = tr + if white { 1 } else { -1 };
            b[capture_row as usize][tc as usize] = ' ';
        }
        if !is_king_in_check(&b, white, castling_rights, None) {
            legal.insert((tc, tr));
        }
    }
    legal
}

/// True if the given side's king is in check.
pub fn is_king_in_check(
    board: &Board,
    white: bool,
    castling_rights: &CastlingRights,
    en_passant_target: Option<(i32, i32)>,
) -> bool {
    let Some((kcol, krow)) = board::find_king(board, white) else {
        return false;
    };
    let opponent_white = !white;
    for r in 0..SQUARES {
        for c in 0..SQUARES {
            if !board::piece_is_side(board, c, r, opponent_white) {
                continue;
            }
            let moves = pseudo_legal_moves(board, c, r, castling_rights, en_passant_target);
            if moves.contains(&(kcol, krow)) {
                return true;
            }
        }
    }
    false
}

/// True if the side to move has no legal moves (checkmate or stalemate).
pub fn no_legal_moves(
    board: &Board,
    white_to_move: bool,
    castling_rights: &CastlingRights,
    en_passant_target: Option<(i32, i32)>,
) -> bool {
    for r in 0..SQUARES {
        for c in 0..SQUARES {
            if !board::piece_is_side(board, c, r, white_to_move) {
                continue;
            }
            if !legal_moves(board, c, r, castling_rights, en_passant_target).is_empty() {
                return false;
            }
        }
    }
    true
}

trait ToAsciiLowercase {
    fn to_ascii_lowercase(self) -> char;
}
impl ToAsciiLowercase for char {
    fn to_ascii_lowercase(self) -> char {
        self.to_lowercase().next().unwrap_or(self)
    }
}

File: src/pieces.py
"""Piece move generation: pseudo-legal moves, legal moves, check detection."""

from .constants import SQUARES
from . import board as board_mod


def _pawn_moves(board, col, row, white, en_passant_target, out):
    forward = -1 if white else 1
    start_row = 6 if white else 1
    r1 = row + forward
    if board_mod.in_bounds(col, r1) and board_mod.is_empty(board, col, r1):
        out.add((col, r1))
        if row == start_row:
            r2 = row + 2 * forward
            if board_mod.in_bounds(col, r2) and board_mod.is_empty(board, col, r2):
                out.add((col, r2))
    for dc in (-1, 1):
        c, r = col + dc, row + forward
        if not board_mod.in_bounds(c, r):
            continue
        p = board_mod.get_piece(board, c, r)
        if p != ' ' and board_mod.is_black_piece(p) == white:
            out.add((c, r))
        elif en_passant_target == (c, r):
            out.add((c, r))


def _sliding_moves(board, col, row, white, directions, out):
    for dc, dr in directions:
        c, r = col + dc, row + dr
        while board_mod.in_bounds(c, r):
            p = board_mod.get_piece(board, c, r)
            if p == ' ':
                out.add((c, r))
            else:
                if board_mod.is_black_piece(p) == white:
                    out.add((c, r))
                break
            c += dc
            r += dr


def _king_moves(board, col, row, white, out):
    for dc in (-1, 0, 1):
        for dr in (-1, 0, 1):
            if dc == 0 and dr == 0:
                continue
            c, r = col + dc, row + dr
            if not board_mod.in_bounds(c, r):
                continue
            p = board_mod.get_piece(board, c, r)
            if p == ' ' or (board_mod.is_black_piece(p) == white):
                out.add((c, r))


def _knight_moves(board, col, row, white, out):
    deltas = [(2, 1), (2, -1), (-2, 1), (-2, -1), (1, 2), (1, -2), (-1, 2), (-1, -2)]
    for dc, dr in deltas:
        c, r = col + dc, row + dr
        if not board_mod.in_bounds(c, r):
            continue
        p = board_mod.get_piece(board, c, r)
        if p == ' ' or (board_mod.is_black_piece(p) == white):
            out.add((c, r))


def _is_square_attacked(board, col, row, attacker_white):
    for r in range(SQUARES):
        for c in range(SQUARES):
            if not board_mod.piece_is_side(board, c, r, attacker_white):
                continue
            piece = board_mod.get_piece(board, c, r)
            lower = piece.lower()
            moves = set()
            if lower == 'p':
                _pawn_moves(board, c, r, attacker_white, None, moves)
            elif lower == 'r':
                _sliding_moves(board, c, r, attacker_white, [(0, 1), (0, -1), (1, 0), (-1, 0)], moves)
            elif lower == 'b':
                _sliding_moves(board, c, r, attacker_white, [(1, 1), (1, -1), (-1, 1), (-1, -1)], moves)
            elif lower == 'q':
                _sliding_moves(board, c, r, attacker_white, [(0, 1), (0, -1), (1, 0), (-1, 0)], moves)
                _sliding_moves(board, c, r, attacker_white, [(1, 1), (1, -1), (-1, 1), (-1, -1)], moves)
            elif lower == 'k':
                _king_moves(board, c, r, attacker_white, moves)
            elif lower == 'n':
                _knight_moves(board, c, r, attacker_white, moves)
            if (col, row) in moves:
                return True
    return False


def _king_castling_moves(board, col, row, white, rights, out):
    if col != 4:
        return
    opponent_white = not white
    if white:
        if row != 7:
            return
        kingside_ok = rights['white_kingside']
        queenside_ok = rights['white_queenside']
    else:
        if row != 0:
            return
        kingside_ok = rights['black_kingside']
        queenside_ok = rights['black_queenside']

    rook_char = 'r' if white else 'R'
    if kingside_ok:
        if (board_mod.is_empty(board, 5, row) and board_mod.is_empty(board, 6, row)
            and board_mod.get_piece(board, 7, row) == rook_char
            and not _is_square_attacked(board, 4, row, opponent_white)
            and not _is_square_attacked(board, 5, row, opponent_white)
            and not _is_square_attacked(board, 6, row, opponent_white)):
            out.add((6, row))
    if queenside_ok:
        if (board_mod.is_empty(board, 1, row) and board_mod.is_empty(board, 2, row)
            and board_mod.is_empty(board, 3, row)
            and board_mod.get_piece(board, 0, row) == rook_char
            and not _is_square_attacked(board, 4, row, opponent_white)
            and not _is_square_attacked(board, 3, row, opponent_white)
            and not _is_square_attacked(board, 2, row, opponent_white)):
            out.add((2, row))


def pseudo_legal_moves(board, col, row, castling_rights, en_passant_target):
    """All pseudo-legal target squares for the piece at (col, row)."""
    piece = board_mod.get_piece(board, col, row)
    if piece == ' ':
        return set()
    white = board_mod.is_white_piece(piece)
    targets = set()
    lower = piece.lower()
    ep = en_passant_target  # (c, r) or None

    if lower == 'p':
        _pawn_moves(board, col, row, white, ep, targets)
    elif lower == 'r':
        _sliding_moves(board, col, row, white, [(0, 1), (0, -1), (1, 0), (-1, 0)], targets)
    elif lower == 'b':
        _sliding_moves(board, col, row, white, [(1, 1), (1, -1), (-1, 1), (-1, -1)], targets)
    elif lower == 'q':
        _sliding_moves(board, col, row, white, [(0, 1), (0, -1), (1, 0), (-1, 0)], targets)
        _sliding_moves(board, col, row, white, [(1, 1), (1, -1), (-1, 1), (-1, -1)], targets)
    elif lower == 'k':
        _king_moves(board, col, row, white, targets)
        _king_castling_moves(board, col, row, white, castling_rights, targets)
    elif lower == 'n':
        _knight_moves(board, col, row, white, targets)
    return targets


def legal_moves(board, col, row, castling_rights, en_passant_target):
    """Filter pseudo-legal moves to those that don't leave own king in check."""
    piece = board_mod.get_piece(board, col, row)
    if piece == ' ':
        return set()
    white = board_mod.is_white_piece(piece)
    pseudo = pseudo_legal_moves(board, col, row, castling_rights, en_passant_target)
    legal = set()
    for tc, tr in pseudo:
        b = [r[:] for r in board]
        b[tr][tc] = piece
        b[row][col] = ' '
        if en_passant_target and (tc, tr) == en_passant_target:
            capture_row = tr + (1 if white else -1)
            b[capture_row][tc] = ' '
        if not is_king_in_check(b, white, castling_rights, None):
            legal.add((tc, tr))
    return legal


def is_king_in_check(board, white, castling_rights, en_passant_target):
    pos = board_mod.find_king(board, white)
    if pos is None:
        return False
    kcol, krow = pos
    opponent_white = not white
    for r in range(SQUARES):
        for c in range(SQUARES):
            if not board_mod.piece_is_side(board, c, r, opponent_white):
                continue
            moves = pseudo_legal_moves(board, c, r, castling_rights, en_passant_target)
            if (kcol, krow) in moves:
                return True
    return False


def no_legal_moves(board, white_to_move, castling_rights, en_passant_target):
    for r in range(SQUARES):
        for c in range(SQUARES):
            if not board_mod.piece_is_side(board, c, r, white_to_move):
                continue
            if legal_moves(board, c, r, castling_rights, en_passant_target):
                return False
    return True
